From 7550c76dc6c0dffe77fb7f5505b1478ad01f0bde Mon Sep 17 00:00:00 2001
From: Piotr Tworek <piotr.tworek@jollamobile.com>
Date: Fri, 3 Jul 2015 14:12:32 +0200
Subject: [PATCH 13/17] Don't decode all images on shell activation if
 decodeondraw is enabled. Fixes JB#25287

Currently if the page contains a lot images in nsDocument::mImageTracker
they will be all decoded when nsDocument::SetImageLockingState(true).
This seems to be an unwanted behaviour when image.mem.decodeondraw pref
is enabled. In such case the images should be decoded just before being
painted. Since some of the images may not be even visible, we don't want to
waste memory for the decoded data.

The original fix for the problem disabled image locking completely.
Unfortunately this cused regressions on pages like jolla.com/tablet (see
JB#30305).
---
 content/base/src/nsDocument.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/content/base/src/nsDocument.cpp b/content/base/src/nsDocument.cpp
index 3493bce..025667c 100644
--- a/content/base/src/nsDocument.cpp
+++ b/content/base/src/nsDocument.cpp
@@ -10020,7 +10020,9 @@ PLDHashOperator LockEnumerator(imgIRequest* aKey,
                                void*    userArg)
 {
   aKey->LockImage();
-  aKey->RequestDecode();
+  if (!Preferences::GetBool("image.mem.decodeondraw", false)) {
+    aKey->RequestDecode();
+  }
   return PL_DHASH_NEXT;
 }
 
-- 
2.3.6

